Cython/Compiler/ModuleNode.py

```diff
@@ -1965,26 +1974,51 @@ def generate_method_table(self, env, code):
         code.putln(
             "};")
 
+    def generate_dict_getter(self, scope, code):
+        #if scope.name == 'QApplication':
+        #    import ipdb;ipdb.set_trace()
+        func_name = scope.mangle_internal("__dict__getter")
+        dict_attr = scope.lookup_here("__dict__")
+        dict_name = dict_attr.cname
+        code.putln("")
+        code.putln("static PyObject *%s(PyObject *o, CYTHON_UNUSED void *x) {" % func_name)
+        self.generate_self_cast(scope, code)
+        code.putln("if (p->%s == 0){" % dict_name)
+        code.putln("p->%s = PyDict_New();" % dict_name)
+        code.putln("}")
+        code.putln("Py_INCREF(p->%s);" % dict_name)
+        code.putln("return p->%s;" % dict_name)
+        code.putln("}")
+
     def generate_getset_table(self, env, code):
-        if env.property_entries:
+        dynamic_attributes = env.lookup_here("__dict__")
+        if env.property_entries or dynamic_attributes:
             code.putln("")
             code.putln(
                 "static struct PyGetSetDef %s[] = {" %
                 env.getset_table_cname)
+            if dynamic_attributes:
+                dict_getter_cname = env.mangle_internal("__dict__getter")
+                code.putln(
+                    '{(char *)"__dict__", %s, 0, 0, 0},' % (
+                        dict_getter_cname))
             for entry in env.property_entries:
-                doc = entry.doc
-                if doc:
-                    if doc.is_unicode:
-                        doc = doc.as_utf8_string()
-                    doc_code = doc.as_c_string_literal()
+                if entry.name == "__dict__":
+                    continue
                 else:
```

Comment:
No need for an "else" after "continue".
And: did you consider merging this feature into the property support to avoid this special casing? (Just asking - I don't say it's a good idea, but it might be.)